keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+
+If truly core APIs are developed as custom resources, then there can be dependency issues when
+bootstrapping a Kubernetes cluster. If core components (e.g. the controller manager) needs to manage
+a custom resource, how do you ensure that the resource gets created before the controller-manager
+starts, when the controller-manager itself is in the lifecycle for custom resources. We need to untangle
+these circular dependencies. And ensure that people who write such controllers, write them in such a
+way that they can temporarily handle the resources they manage not existing.
+
+Another significant concern in dependency management is dependencies between custom resource definitions.
+As the ecosystem of plugins becomes richer, the chances that one plugin will depend on another increases.
+Currently there is no good way to express that in order for CustomResource `A` to work correctly, 
+CustomResource `B` needs to be installed as well.
+
+#### Version upgrades
+
+Another open question for CustomResources is version management. Currently there are no built-in practices
+for moving from one version of a CustomResource to another. To do such a migration, the tool itself would
+have to perform the schema migration by reading all of the data objects in the old version and rewriting
+them in the new version. To make this successful , at the very least we will need to provide the ability to
+mark a CustomResource as "read-only" to ensure that there isn't skew between the two representations.
+
+We may need to invest significantly more heavily in this space to make it easy for users to upgrade or 
+update their objects. For now, most users have simply chosen not to upgrade the version.
+
+### Risks and Mitigations
+
+#### DDOS or other attacks on storage
+
+Because custom resources share the same storage as the core API types, there is risk that either via 
+requests/second or storage space usage, custom resources will become deliberate or inadvertent denial
+of service attacks on a cluster. Currently there is a maximum size cap for custom resources, but it is
+a pretty limited defense against these sorts of concerns. This makes the installation of a custom
+resource a very privileged action for a cluster.
+
+#### Custom resource adminstration
+
+For many reasons, installing a new custom resource is a privileged action for a cluster adminstrator.
+However, this significantly increases the complexity of developing a Custom Resource. Ideally the
+resource itself would install itself into the cluster when it's controller is run as a Pod, but
+because you don't want to grant the controller the general ability to create arbitrary CustomResources,
+it is generally the case that the controller won't have proper permissions to create a resource. 
+
+In the future we may want to develop some sort of "single use token" or some equivalent such that custom resources are easy to install, and yet don't require cluster adminstrator privileges to bootstrap themselves.
```

Comment:
We definitely have the goal of making them easy, but not necessarily in that way. The fundamental problem isn't that it's hard to give permissions (that's easy to fix), it's that you shouldn't give permissions to folks who lack the full context of what the cluster is used for. :(

Giving folks their own sandboxed apiserver or section of apiserver is one approach that's been discussed on the mailing lists, but it's a lot of work.

Reply:
Can anyone have a full context of what a cluster is used for? A cluster admin may know about Kubernetes but they aren't likely to understand all the apps, controllers, and other things running in a cluster along with what they are doing. In a cluster of a moderate size or in an organization with various loosely connected teams using it, can anyone know the full context?

Reply:
> Can anyone have a full context of what a cluster is used for?

It is reasonable for some clusters, less for others.

The harder it is to know the full context, the more important it is to be careful about tweaking global bits (e.g., installing CRDs).